containerd integration: add support for since,before,labels images filters

thaJeztah opened the pull request:
- depends on / follow-up to https://github.com/moby/moby/pull/43815
- upstreaming https://github.com/rumpl/moby/pull/7


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

ndeloof committed:
add support for since, before, labels images filters

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Rebased, because https://github.com/moby/moby/pull/43815 was merged

tianon reviewed (approved):
I was looking at https://github.com/moby/moby/blob/2fbc30739b437d91f7a444f49143305e5f0a303e/daemon/images/image_list.go hoping we could have some overlapping implementation here but saw quickly that the filtering is integral to the list generation itself over there, so I guess not so much. :sweat_smile:

I do wonder (as a user of many systems with an extremely prolific number of images on them) how well this implementation will scale -- I guess the other way to put that is that this is probably somewhere the `containerd` API perhaps needs some changes for batching/streaming/"iterator" of the image list instead of returning the full list of objects in a single call. :innocent: :see_no_evil:

thaJeztah commented:
> I guess the other way to put that is that this is probably somewhere the containerd API perhaps needs some changes for batching/streaming/"iterator" of the image list instead of returning the full list of objects in a single call. 😇 🙈

Yes, I think we should look if this can be implemented in upstream containerd somehow; I agree that the current approach is not ideal (sending a lot of info over the wire, and filtering after the fact)

thaJeztah commented:
@corhere @cpuguy83 PTAL

cpuguy83 commented on daemon/containerd/image_list.go:
```diff
@@ -5,21 +5,36 @@ import (
+
 // Images returns a filtered list of images.
 func (i *ImageService) Images(ctx context.Context, opts types.ImageListOptions) ([]*types.ImageSummary, error) {
 	imgs, err := i.client.ListImages(ctx)
```
You can probably do the filters as fieldpaths in this call.

cpuguy83 replied:
https://github.com/containerd/containerd/blob/main/filters/filter.go

Doing the date comparison is not possible today, but most things should be.

thaJeztah replied:
Oh, that looks interesting indeed. Will have to dig into that a bit to see how to integrate it (I'm mostly upstreaming patches here 😂)

what's your preference? trying to make that work before we merge this, or improve in a follow-up?

cpuguy83 replied:
I'm fine with it as is, let's maybe make sure to note though that the implementation needs work.

thaJeztah replied:
I added some todo's and tracking issues;

- https://github.com/moby/moby/issues/43845
- https://github.com/moby/moby/issues/43846
- https://github.com/moby/moby/issues/43847
- https://github.com/moby/moby/issues/43848

corhere commented on daemon/containerd/image_list.go:
```diff
@@ -46,6 +61,54 @@ func (i *ImageService) Images(ctx context.Context, opts types.ImageListOptions)
 	return ret, nil
 }
 
+func (i *ImageService) setupFilters(ctx context.Context, opts types.ImageListOptions) ([]imageFilterFunc, error) {
```
```suggestion
func (i *ImageService) setupFilters(ctx context.Context, opts types.ImageListOptions) (imageFilterFunc, error) {
```
```go
return func(image containerd.Image) bool {
        for _, filter := range filters {
                if !filter(image) {
                    return false
                }
        }
        return true
}, nil
```

There is only one filter. It may have more than one clause, but they're all logically parts of one filter. The caller of `setupFilters` shouldn't need to care how the filter is implemented; returning a slice of filter clauses just opens up opportunities for the calling code to evaluate the filter incorrectly (e.g. by ORing the clauses when they should be ANDed).

thaJeztah replied:
Good suggestion; I was also looking at that.

- I updated the function as suggested
- Changed the signature to accept `filters.Args` instead of the whole `ImageListOptions`
- I moved the `i.setupFilters()` invocation to the start of the function, as parsing the filters should be less work than first fetching all images (only to return an error if there's an issue with constructing the filters)
- I added a second commit to validate the filters (as not all filters are implemented yet)
- Created some tracking issues for that; see  https://github.com/moby/moby/pull/43816#discussion_r926449121

thaJeztah committed:
daemon/containerd: validate image filters

Not all filters are implemented yet, so make sure an error
is returned if a not-yet implemented filter is used.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Interesting failure on Windows (could be just a hiccup) https://github.com/moby/moby/runs/7446210210?check_suite_focus=true

```
Install-Package : No match was found for the specified search criteria and 
package name '7Zip4PowerShell'. Try Get-PackageSource to see all available 
registered package sources.
At line:1 char:3109
+ ... ownloading containerd; Install-Package -Force 7Zip4PowerShell; $locat ...
+                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Microsoft.Power....InstallPacka 
   ge:InstallPackage) [Install-Package], Exception
    + FullyQualifiedErrorId : NoMatchFoundForCriteria,Microsoft.PowerShell.Pac 
   kageManagement.Cmdlets.InstallPackage
```

thaJeztah commented:
I'm gonna ignore the `TestRunInteractiveWithRestartPolicy` failure on windows as we know it's unrelated and flaky
